synapse/storage/databases/main/events.py

```diff
@@ -993,6 +1004,11 @@ def _update_current_state_txn(
                 """
                 txn.execute(sql, (stream_id, self._instance_name, room_id))
 
+                # We also want to invalidate the membership caches for remote
+                # users that were in the room.
+                users_in_room = self.store.get_users_in_room_txn(txn, room_id)
+                members_changed.update(users_in_room)
```

Comment:
I might just be me, but I'm finding the remote-ness discussion confusing.

I think what we want to do is:

1. If the server has left the room, `members_changed` should be all users in the room (so `get_rooms_for_user` gets invalidated for everyone).
2. If the server is still in the room, `members_changed` should be calculated from the state changes to the room.

Is that accurate or no?

Reply:
Yup, that is correct. The thing is that `to_delete` will already include all local users in the room (as otherwise the server won't have "left" the room). So the impact of this bug is for remote users only, but I'll update the comment to clarify